File: recursion/javascript/solutions/largest.js
/**
 * Solution for largest.
 */

const { Node, isEmpty, unprepend } = require('./linkedList');

function largest(list) {
  if (isEmpty(list.next)) {
    return list.value;
  }

  let [first, rest] = unprepend(list);

  return Math.max(first, largest(rest));
}

if (require.main === module) {
  console.assert(largest(Node.fromArray([3, 7, 2])) === 7);
  console.assert(largest(Node.fromArray([1, 2, 3, 4, 5])) === 5);
  console.assert(largest(Node.fromArray([5, 4, 3, 2, 1])) === 5);
  console.assert(largest(Node.fromArray([42])) === 42);
  console.assert(largest(Node.fromArray([-5, -3, -10])) === -3);

  console.log('All tests passed!');
}

module.exports = { largest };

File: recursion/python/solutions/largest.py
"""Solution for largest."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend


def largest(lst):
    """
    Given a non-empty list of numbers, return the largest.

    largest(3 -> 7 -> 2) => 7

    Note: You can assume the list is non-empty.
    """
    # Base case: single element list
    if is_empty(lst.next):
        return lst.value

    first, rest = unprepend(lst)

    return max(first, largest(rest))


if __name__ == '__main__':
    # Test cases
    assert largest(Node.from_list([3, 7, 2])) == 7
    assert largest(Node.from_list([1, 2, 3, 4, 5])) == 5
    assert largest(Node.from_list([5, 4, 3, 2, 1])) == 5
    assert largest(Node.from_list([42])) == 42
    assert largest(Node.from_list([-1, -5, -2])) == -1
    assert largest(Node.from_list([0, 0, 0])) == 0

    print("All tests passed!")
